src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -7157,6 +7154,12 @@ private BoundExpression TryImplicitConversionToArrayIndex(BoundExpression expr,
             var result = TryImplicitConversionToArrayIndex(expr, type, node, attemptDiagnostics);
             if (result is object)
             {
+                if (!IsSemanticModelBinder)
```

Comment:
>IsSemanticModelBinder [](start = 21, length = 21)

We're checking/passing this condition in a bunch of different places. Consider adding a `RecordUsedAssemblyIfNecessary` on `Binder` that checks this in one place.

Reply:
> We're checking/passing this condition in a bunch of different places. Consider adding a RecordUsedAssemblyIfNecessary on Binder that checks this in one place.

I am not sure I am following. It looks like there are only two ```if``` statements like this in binder code. I don't believe it is worth adding a helper for that.

---
In reply to: [351359049](https://github.com/dotnet/roslyn/pull/40010#discussion_r351359049) [](ancestors = 351359049)

Reply:
> It looks like there are only two if statements like this in binder code

The previous PR had 2 direct copies of this, plus 2 where this was checked in a nested local function (`Binder.ResultSymbol`). The symbol code also had a few cases where the symbol checked the property off the binder it was passed.

---
In reply to: [351437499](https://github.com/dotnet/roslyn/pull/40010#discussion_r351437499) [](ancestors = 351437499,351359049)

Reply:
And as we add tracking more scenarios, I imagine that will continue to expand.

---
In reply to: [351445110](https://github.com/dotnet/roslyn/pull/40010#discussion_r351445110) [](ancestors = 351445110,351437499,351359049)

Reply:
> The previous PR had 2 direct copies of this, plus 2 where this was checked in a nested local function (Binder.ResultSymbol). The symbol code also had a few cases where the symbol checked the property off the binder it was passed.

I do not see this. There are only two places where we do ```if (!IsSemanticModelBinder) Compilation.AddUsedAssembly(...)``` and all of them are added in this PR. Binder.ResultSymbol doesn't even call ```Compilation.AddUsedAssembly```. It is quite possible I still do not understand what exact refactoring you are suggesting. Are you suggesting to add another property to return ```!IsSemanticModelBinder``` from it and use it instead of checking ```IsSemanticModelBinder``` inline? 

> And as we add tracking more scenarios, I imagine that will continue to expand.

I do not plan to add any more similar tracking during binding.

---
In reply to: [351445297](https://github.com/dotnet/roslyn/pull/40010#discussion_r351445297) [](ancestors = 351445297,351445110,351437499,351359049)

Reply:
[https://github.com/dotnet/roslyn/commit/c5bc1e029b117859a79cdc505ad46974005f8988#diff-a921499c9550563c162f3e1cbdff548bR387-R390](https://github.com/dotnet/roslyn/commit/c5bc1e029b117859a79cdc505ad46974005f8988#diff-a921499c9550563c162f3e1cbdff548bR387-R390)
[https://github.com/dotnet/roslyn/commit/c5bc1e029b117859a79cdc505ad46974005f8988#diff-56a949e9d40faeb9e94d55a560b7a6e5R197-R200](https://github.com/dotnet/roslyn/commit/c5bc1e029b117859a79cdc505ad46974005f8988#diff-56a949e9d40faeb9e94d55a560b7a6e5R197-R200)

These aren't quite the same, but calling a very similar API with the same condition:
[https://github.com/dotnet/roslyn/commit/c5bc1e029b117859a79cdc505ad46974005f8988#diff-591e192a1c591a08bec434a1c6a654abR318-R322](https://github.com/dotnet/roslyn/commit/c5bc1e029b117859a79cdc505ad46974005f8988#diff-591e192a1c591a08bec434a1c6a654abR318-R322)
[https://github.com/dotnet/roslyn/commit/c5bc1e029b117859a79cdc505ad46974005f8988#diff-591e192a1c591a08bec434a1c6a654abR335-R340](https://github.com/dotnet/roslyn/commit/c5bc1e029b117859a79cdc505ad46974005f8988#diff-591e192a1c591a08bec434a1c6a654abR335-R340)
[https://github.com/dotnet/roslyn/commit/c5bc1e029b117859a79cdc505ad46974005f8988#diff-3d0fdd05badb266fa634cddaee906ddeR197-R201](https://github.com/dotnet/roslyn/commit/c5bc1e029b117859a79cdc505ad46974005f8988#diff-3d0fdd05badb266fa634cddaee906ddeR197-R201)

---
In reply to: [351451006](https://github.com/dotnet/roslyn/pull/40010#discussion_r351451006) [](ancestors = 351451006,351445297,351445110,351437499,351359049)

Reply:
The point being that, in general, I don't really think that `IsSpeculativeSemanticModel` is the type of check that it would occur to me that I need to perform before recording whether or not I'm using a type reference. I'm worried that in the future, someone will add code here that just adds the used assembly unconditionally, so it would be better to have that check made for you as part of a `RecordIfNecessary` API.

---
In reply to: [351453835](https://github.com/dotnet/roslyn/pull/40010#discussion_r351453835) [](ancestors = 351453835,351451006,351445297,351445110,351437499,351359049)

Reply:
> so it would be better to have that check made for you as part of a RecordIfNecessary API.

Sorry, I do not see the benefit as anyone can call APIs on compilation directly. Again, it is quite possible that I am misinterpreting your suggestion. Perhaps you could provide the code for what it it you are suggesting.

---
In reply to: [351455345](https://github.com/dotnet/roslyn/pull/40010#discussion_r351455345) [](ancestors = 351455345,351453835,351451006,351445297,351445110,351437499,351359049)

Reply:
Perhaps we can follow up on this later in-person?

---
In reply to: [351457798](https://github.com/dotnet/roslyn/pull/40010#discussion_r351457798) [](ancestors = 351457798,351455345,351453835,351451006,351445297,351445110,351437499,351359049)